### DISCUSSION TITLE
Development

### DISCUSSION THREAD

**Discussion by Chreece** [Chreece/BTHomeV2-ESP32-example#12] [General]
Continuing the discussion from [here](https://github.com/Chreece/BTHomeV2-ESP32-example/issues/3#issuecomment-1546682448):

Yes every entity is showing up in HA

**Comment by Chreece** [comment]
So my view on how the code should work is:

In .ino file the user only inputs device_name, encryption & bind_key and the object_ids + values.
These values should be send at once in a method in .cpp file to handle them.

The method should check after new calling if the data could fit in one packet and send them all at once, or chop them and send them seperately.

So the setup() in .ino should be something like:
Send to the method the device_name and encryption with bind_key if encryption.

The loop():
Send the object_ids + values seperately for every sensor.
delay + sleep

Then the method in .cpp can judge what to do...

That will be my final goal for this project and can sleep hapilly :)

**Reply by countrysideboy** [reply]
In facts, I have add some code to compare the data length before. Perhaps the most complex part is how to achieve the nonblocking function.

**Reply by Chreece** [reply]
I saw it and I'm trying to use it

**Comment by countrysideboy** [comment]
These days, I try to use ble 5.0 features, such as long range and ext advertising.As I move the device name to the scan response packet, the old ble4.0 can meet my need.

**Comment by Chreece** [comment]
If you have time check the DEV branch, i adjusted the code to chop the packets, it doesn't work perfectly for every sensor, maybe you can find some errors

**Reply by countrysideboy** [reply]
According to your dev branch, I'm coding another version that can chop and send the packet. 

My final goal is non-blocking code.

**Reply by countrysideboy** [reply]
I just submitted a very rough code, but I have no good idea about non-blocking right now.
I think it shoud be something like that without delay blocking:

loop()
{
if(bthome.run())
{
add...
add...
}
}

**Comment by Chreece** [comment]
I'm no programmer, don't know the term non blocking 😜
